File: haskell/bob/src/Bob.hs
module Bob (responseFor) where

import Data.Char
import Data.List

responseFor :: String -> String
responseFor input
	| isNothing input =  "Fine. Be that way!"
	| isYell input = "Whoa, chill out!"
	| isQuestion input = "Sure."
	| otherwise = "Whatever."

isNothing :: String -> Bool
isNothing =
	(==) "" . dropWhile isSpace

isYell :: String -> Bool
isYell string =
	let
		isSame =
			map toUpper string == string
	in
		hasLetters string && isSame

isQuestion :: String -> Bool
isQuestion =
	(==) '?' . last . removeWhitespace

letters :: String
letters =
	['a'..'z'] ++ ['A'..'Z']

hasLetters :: String -> Bool
hasLetters =
	any id . map (\c -> elem c letters)

removeWhitespace :: String -> String
removeWhitespace =
	filter (\c -> c /= ' ')

File: nim/bob/bob.nim
import nre 
import strutils

proc isNothing(input: string): bool =
  strip(input) == ""

proc isYell(input: string): bool =
  input == toUpper(input) and contains(input, re"""[A-Z]""")

proc isQuestion(input: string): bool =
  endsWith(strip(input), "?")

proc hey*(input: string): string =
  if isNothing(input):
    "Fine. Be that way!"
  elif isYell(input):
    "Whoa, chill out!"
  elif isQuestion(input):
    "Sure."
  else:
    "Whatever."

